CI: Update docs-related workflows

akihironitta opened the pull request:
## What does this PR do?
- Closes #13422.
- Updates documentation-related CI/CD.

<!---
OUTDATED
- Fixes the current unavailability of CircleCI job building docs.
- Removes CI docs build in favour of [#13572](https://github.com/Lightning-AI/lightning/issues/13572):
  - from `.github/workflows/docs-check.yml`, and
  - from `.circleci/config.yml`
--->


refs:
- https://hub.docker.com/r/readthedocs/build
- https://github.com/readthedocs/readthedocs-docker-images
- https://readthedocs.org/projects/pytorch-lightning/builds/17356368/

### Does your PR introduce any breaking changes? If yes, please list them.
None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [n/a] Did you make sure to **update the documentation** with your changes? (if necessary)
- [n/a] Did you write any **new necessary tests**? (not for typos and docs)
- [n/a] Did you verify new and **existing tests pass** locally with your changes?
- [n/a] Did you list all the **breaking changes** introduced by this pull request?
- [n/a] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃


cc @carmocca @akihironitta @borda

Borda reviewed (commented):
I would prefer to keep using official readthedocs image to flag some potential issues, having the same as gha does not have much sense...

akihironitta commented:
> I would prefer to keep using official readthedocs image to flag some potential issues

Sounds reasonable!

> having the same as gha does not have much sense...

I wonder why there's basically the same CI job among our GH workflows as this CircleCI job. Do you mean `.github/workflows/docs-checks.yml` can be completely removed (which I think can be)?

Borda commented:
> I wonder why there's basically the same CI job among our GH workflows as this CircleCI job.

CircleCI offer inspecting artifacts online which means you can browse built docs inline without downloading them and opening them locally... :rabbit:

akihironitta commented:
> CircleCI offer inspecting artifacts online which means you can browse built docs inline without downloading them and opening them locally... 🐰

I know, and we should definitely keep the CircleCI job for easy reference to built docs, but I'm asking if we can remove the GH workflow `.github/workflows/docs-checks.yml`. What do you think?

akihironitta commented on .github/workflows/docs-checks.yml:
```diff
@@ -1,109 +0,0 @@
-          make doctest
-          make coverage
-
-  make-docs:
```
Removing `make-docs` since this is a complete duplicate of the CircleCI job building docs.

Borda replied:
tbh, if we find a way how to share/browse artefact from GHA, I would rather drop CircleCI and keep in-house

akihironitta replied:
> I would rather drop CircleCI 

@Borda Found a possible alternative #13572 to CircleCI :]

akihironitta replied:
Resolving this comment as the removal of doctest has been reverted.

akihironitta commented on .github/workflows/docs-checks.yml:
```diff
@@ -1,109 +0,0 @@
-  cancel-in-progress: ${{ ! (github.ref == 'refs/heads/master' || startsWith(github.ref, 'refs/heads/release/')) }}
-
-jobs:
-  doctest:
```
Removing `doctest` since we do have our test at https://github.com/Lightning-AI/lightning/blob/61c28cb428a13c2aea6d7f3f55e0f00431a4ea4e/.github/workflows/ci-pytorch_test-full.yml#L95

Borda replied:
it is very different and just name colisiotn, this doctest is running python code in `.rst` files, compare to doctest you refer is running examples un docstring in `.py` files

akihironitta replied:
Resolving this comment as the removal of doctest has been reverted.

akihironitta commented on .circleci/config.yml:
```diff
@@ -140,9 +144,11 @@ jobs:
   build-Docs:
     docker:
-      - image: readthedocs/build:latest
+      - image: readthedocs/build:ubuntu-20.04-2022.02.16
```
`readthedocs/build:latest` is two years old (see [here](https://hub.docker.com/layers/build/readthedocs/build/latest/images/sha256-3bc8a04cafd51ca7520b1c0c4b002f6aaac52695fa4d9f69c3ee92fac689678c?context=explore)), and I'm quite sure that it isn't used by readthedocs.

Borda replied:
very nice finding! :rabbit:

Borda commented on .github/workflows/docs-checks.yml:
```diff
@@ -1,109 +0,0 @@
-        working-directory: ./docs
-        run: |
-          # First run the same pipeline as Read-The-Docs
-          make html --debug --jobs $(nproc) SPHINXOPTS="-W --keep-going"
```
let's keep it with calling `make docs` from root

akihironitta replied:
Why would it be necessary? I don't think we need to build docs in any of GHA workflows if we follow #13572.

Borda replied:
maybe not build but doctest it :) https://github.com/Lightning-AI/lightning/issues/13572#issuecomment-1185727378

akihironitta commented on .github/workflows/docs-checks.yml:
```diff
@@ -1,109 +0,0 @@
-        working-directory: ./docs
-        run: |
-          # First run the same pipeline as Read-The-Docs
-          make doctest
```
I think we already run doctest with `pytest` in another place https://github.com/Lightning-AI/lightning/pull/13547#discussion_r915436509

akihironitta replied:
Resolving this comment as it's being discussed in https://github.com/Lightning-AI/lightning/pull/13547#discussion_r915436509.

akihironitta commented on docs/Makefile:
```diff
@@ -2,7 +2,7 @@
 
 # You can set these variables from the command line.
-SPHINXOPTS    =
+SPHINXOPTS    = -T -W --keep-going
```
```
  -W                turn warnings into errors
  --keep-going      with -W, keep going when getting warnings
  -T                show full traceback on exception
```

Borda reviewed (changes requested):
we need to keep the GHA jobs as we need to build two independent docs instances which won't be possible to parameterize with pure readthedocs, ref #13537

akihironitta commented:
> we need to keep the GHA jobs as we need to build two independent docs instances which won't be possible to parameterize with pure readthedocs, ref #13537

@Borda Sorry I'm not following the CI/CD on app side. I'm still not sure of the need even after I had a look at the PR (maybe because I know nothing about CI/CD in the app repo, e.g., how the app documentation is hosted).

However, I'm now reverting the change to the GHA as you suggest!

akihironitta commented:
So far, I fixed the failure of building docs in CircleCI at the time of 6fae278. Then, given the review comment https://github.com/Lightning-AI/lightning/pull/13547#discussion_r915523547, I removed the CircleCI job and suggested its alternative #13572.

Does #13572 look like a good alternative? If yes, this PR is blocked by #13572. Otherwise, I'm not sure what to do from here. Do we want to keep running the CircleCI job or looking for other alternatives? It would be great if anyone could take over this PR (or create another PR) to fix the current CircleCI failure in building docs.
